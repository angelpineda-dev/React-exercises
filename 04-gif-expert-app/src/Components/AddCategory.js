import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  // Function to Input
  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  // Function to Form
  const handleSubmit = (e) => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setInputValue("");
    } else {
      console.warn("El campo no debe estar vacio o ser mayor a una letra");
    }
  };

  // What this object return?
  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Buscar..."
      />
    </form>
  );
};

// This object is required...
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
