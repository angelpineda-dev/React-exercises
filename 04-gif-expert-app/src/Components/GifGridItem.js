import React from "react";
import PropTypes from "prop-types";

export const GifGridItem = ({ title, url }) => {
  return (
    <div className="card animate__animated animate__fadeIn">
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};

/* ¿Cómo implementar propTypes para requerir valores?
 - Las PropTypes son utiles para indicar que clase de valor debe tener una variable y más propiedades.
*/

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
