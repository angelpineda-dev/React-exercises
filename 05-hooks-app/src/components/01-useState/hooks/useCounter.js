import { useState } from "react";

/* Custom hook no son mÃ¡s que simples funciones */
export const useCounter = (initialState = 1) => {
  const [counter, setCounter] = useState(initialState);

  const increment = () => {
    setCounter(counter + 1);
  };

  const decrement = () => {
    setCounter(counter - 1);
  };

  const reset = () => {
    setCounter(initialState);
  };

  return {
    counter,
    increment,
    decrement,
    reset,
  };
};
