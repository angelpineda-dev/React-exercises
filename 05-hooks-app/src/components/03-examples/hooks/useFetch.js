import { useState, useEffect, useRef } from "react";

export const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);

  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        setTimeout(() => {
          if (isMounted) {
            setState({
              loading: false,
              error: null,
              data,
            });
          }
        }, 1);
      });
  }, [url]);

  return state;
};
