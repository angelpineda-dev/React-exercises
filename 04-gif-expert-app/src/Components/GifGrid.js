import React /* { useState, useEffect } */ from "react";
import PropTypes from "prop-types";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  // making a request from our hook.
  const { data: images, loading } = useFetchGifs(category);

  return (
    // We're returning a fragment
    <>
      <h3>{category}</h3>

      {loading && (
        <p className="animate__animated animate__flash">Loading...</p>
      )}

      <div className="card-flex">
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
