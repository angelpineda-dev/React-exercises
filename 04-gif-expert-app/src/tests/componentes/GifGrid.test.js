import React from "react";
import { GifGrid } from "../../Components/GifGrid";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { useFetchGifs } from "../../hooks/useFetchGifs";
jest.mock("../../hooks/useFetchGifs");

describe("Pruebas en <GifGrid />", () => {
  let category = "goku";

  test("Debe mostrar correctamente el componente.", () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true,
    });

    let wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe mostrar items cuando se cargan imÃ¡gnes usando useFetchGifs.", () => {
    const gifs = [
      {
        id: "ABC",
        title: "algun titulo",
        url: "https://localhost.com/algo/img.jpg",
      },
      {
        id: "123",
        title: "algun titulo",
        url: "https://localhost.com/algo/img.jpg",
      },
    ];

    useFetchGifs.mockReturnValue({
      data: gifs,
      loading: false,
    });

    let wrapper = shallow(<GifGrid category={category} />);
    //expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
