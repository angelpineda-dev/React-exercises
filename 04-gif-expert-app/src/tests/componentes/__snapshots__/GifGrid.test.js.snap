// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGrid /> Debe mostrar correctamente el componente. 1`] = `
<Fragment>
  <h3>
    goku
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-flex"
  />
</Fragment>
`;
